src/swap/defi/lifi.ts

```diff
@@ -0,0 +1,392 @@
+//   fromToken: asToken,
+//   toChainId: asNumber,
+//   toToken: asToken
+// })
+
+// const asFeeCost = asObject({
+//   amount: asNumberString, // "56495962827064236208",
+//   token: asToken
+// })
+
+const asEstimate = asObject({
+  fromAmount: asNumberString, // "400000",
+  toAmount: asNumberString, // "237318132569913",
+  toAmountMin: asNumberString, // "225452225941418",
+  approvalAddress: asString, // "0x1231DEB6f5749EF6cE6943a275A1D3E7486F4EaE",
+  executionDuration: asNumber // 1168,
+  // feeCosts: asArray(asFeeCost)
+})
+
+const asTransactionRequest = asObject({
+  data: asString,
+  to: asString, // '0x1231DEB6f5749EF6cE6943a275A1D3E7486F4EaE'
+  value: asString, // '0x00'
+  from: asString, // '0x0b0901e9cef9eaed5753519177e3c7cfd0ef96ef'
+  chainId: asNumber, // 1
+  gasPrice: asString, // '0x03aca2109d'
+  gasLimit: asString // '0x08a3df'
+})
+
+const asIncludedStep = asObject({
+  toolDetails: asObject({
+    name: asString
+  })
+})
+
+const asV1Quote = asObject({
+  id: asString,
+  type: asString,
+  estimate: asEstimate,
+  includedSteps: asArray(asIncludedStep),
+  // action: asAction,
+  transactionRequest: asTransactionRequest
+})
+
+// type AssetSpread = ReturnType<typeof asAssetSpread>
+type ExchangeInfo = ReturnType<typeof asExchangeInfo>
+
+let exchangeInfo: ExchangeInfo | undefined
+let exchangeInfoLastUpdate: number = 0
+
+export function makeLifiPlugin(opts: EdgeCorePluginOptions): EdgeSwapPlugin {
+  const { io, log } = opts
+  const { affiliateFeeBasis } = asInitOptions(opts.initOptions)
+  const affiliateFee = div18(affiliateFeeBasis, '10000')
+
+  const headers = {
+    'Content-Type': 'application/json'
+  }
+
+  const fetchSwapQuoteInner = async (
+    request: EdgeSwapRequestPlugin
+  ): Promise<SwapOrder> => {
+    const {
+      fromCurrencyCode,
+      fromTokenId,
+      toCurrencyCode,
+      toTokenId,
+      nativeAmount,
+      fromWallet,
+      toWallet,
+      quoteFor
+    } = request
+    if (quoteFor !== 'from') {
+      throw new SwapCurrencyError(swapInfo, fromCurrencyCode, toCurrencyCode)
+    }
+
+    const fromToken = fromWallet.currencyConfig.allTokens[fromTokenId ?? '']
+    let fromContractAddress
+    let sendingToken = false
+    if (fromCurrencyCode === fromWallet.currencyInfo.currencyCode) {
+      fromContractAddress = fromCurrencyCode
+    } else {
+      sendingToken = true
+      fromContractAddress = fromToken?.networkLocation?.contractAddress
+    }
+
+    const toToken = toWallet.currencyConfig.allTokens[toTokenId ?? '']
+    let toContractAddress
+    if (toCurrencyCode === toWallet.currencyInfo.currencyCode) {
+      toContractAddress = toCurrencyCode
+    } else {
+      toContractAddress = toToken?.networkLocation?.contractAddress
+    }
+
+    if (fromContractAddress == null || toContractAddress == null) {
+      throw new SwapCurrencyError(swapInfo, fromCurrencyCode, toCurrencyCode)
+    }
+
+    const { appId, integrator } = asInitOptions(opts.initOptions)
```

Comment:
Nit pick, these consts and the lifiServers declaration should be after all of the early return scenarios (next to the now declaration)